ambry-messageformat/src/main/java/com.github.ambry.messageformat/MessageFormatSend.java

```diff
@@ -110,44 +111,47 @@ private void calculateOffsets(boolean enableDataPrefetch) throws IOException, Me
             readSet.doPrefetch(i, 0, readSet.sizeInBytes(i));
           }
         } else {
-          // read header version
           long startTime = SystemTime.getInstance().milliseconds();
-          ByteBuffer headerVersion = ByteBuffer.allocate(Version_Field_Size_In_Bytes);
-          readSet.writeTo(i, Channels.newChannel(new ByteBufferOutputStream(headerVersion)), 0,
-              Version_Field_Size_In_Bytes);
-          logger.trace("Calculate offsets, read header version time: {}",
+          BufferedInputStream bufferedInputStream =
+              new BufferedInputStream(new MessageReadSetIndexInputStream(readSet, i, 0), 1024);
+          logger.trace("Calculate offsets, BufferedInputStream initialization time: {}",
               SystemTime.getInstance().milliseconds() - startTime);
 
-          headerVersion.flip();
-          short version = headerVersion.getShort();
+          startTime = SystemTime.getInstance().milliseconds();
+          byte[] headerVersionBytes = new byte[Version_Field_Size_In_Bytes];
+          bufferedInputStream.read(headerVersionBytes, 0, Version_Field_Size_In_Bytes);
+          short version = ByteBuffer.wrap(headerVersionBytes).getShort();
           if (!isValidHeaderVersion(version)) {
             throw new MessageFormatException(
                 "Version not known while reading message - version " + version + ", StoreKey " + readSet.getKeyAt(i),
                 MessageFormatErrorCodes.Unknown_Format_Version);
           }
-          ByteBuffer header = ByteBuffer.allocate(getHeaderSizeForVersion(version));
-          // read the header
+          logger.trace("Calculate offsets, read and verify header version time: {}",
+              SystemTime.getInstance().milliseconds() - startTime);
+
           startTime = SystemTime.getInstance().milliseconds();
-          headerVersion.clear();
-          header.putShort(headerVersion.getShort());
-          readSet.writeTo(i, Channels.newChannel(new ByteBufferOutputStream(header)), Version_Field_Size_In_Bytes,
-              header.capacity() - Version_Field_Size_In_Bytes);
+          byte[] headerBytes = new byte[getHeaderSizeForVersion(version)];
+          bufferedInputStream.read(headerBytes, Version_Field_Size_In_Bytes,
+              headerBytes.length - Version_Field_Size_In_Bytes);
           logger.trace("Calculate offsets, read header time: {}", SystemTime.getInstance().milliseconds() - startTime);
 
           startTime = SystemTime.getInstance().milliseconds();
-          header.flip();
+          ByteBuffer header = ByteBuffer.wrap(headerBytes);
+          header.putShort(version);
+          header.clear();
```

Comment:
`rewind()` is what you really want here and is more readable (though `clear()` has the same effect in this case).

Reply:
`rewind()` doesn't work here. After `putShort()`, `limit` is 2, but we want to have a `limit` equals to `capacity`.

Reply:
`putShort()` (or any write call) does not update the limit, only the position. Since you are not doing a `flip()`, the limit will remain the same (which is initialized to the capacity when you do a `wrap()`). I think `rewind()` is exactly what you want here.

Reply:
nice. Will use `rewind` here.